addons/payment/models/account_payment.py

```diff
@@ -110,30 +110,26 @@ def _prepare_payment_transaction_vals(self, **extra_create_values):
         }
 
     def action_post(self):
-        # Post the payments "normally" if no transactions are needed.
-        # If not, let the acquirer update the state.
-
-        payments_need_tx = self.filtered(
-            lambda p: p.payment_token_id and not p.payment_transaction_id
-        )
-        # creating the transaction require to access data on payment acquirers, not always accessible to users
-        # able to create payments
-        transactions = payments_need_tx.sudo()._create_payment_transaction()
-
-        res = super(AccountPayment, self - payments_need_tx).action_post()
-
-        for tx in transactions:  # Process the transactions with a payment by token
-            tx._send_payment_request()
-
-        # Post payments for issued transactions
+        # Create the missing payment transactions.
+        payments_need_trans = self.filtered(lambda pay: pay.payment_token_id and not pay.payment_transaction_id)
+        payments_need_trans._create_payment_transaction()
+
+        # Process payment transactions directly.
+        payments_with_trans = self.filtered('payment_transaction_id')
+        transactions = payments_with_trans.payment_transaction_id
+        transactions._send_payment_request()
+
+        # Post payments.
+        payments_to_post = self.filtered(lambda pay: not pay.payment_transaction_id
+                                                     or pay.payment_transaction_id.state == 'done')
+        res = super(AccountPayment, payments_to_post).action_post()
+
+        # Post process transactions.
+        transactions = payments_need_trans.payment_transaction_id.filtered(lambda x: x.state == 'done')
```

Comment:
It's bad practice to overwrite an already defined variable. In particular when there are so many `filtered` in play :|

You could simply save the recordset returned by `_create_payment_transaction` into `transactions_sudo` and filter it based on `state`.